### ISSUE TITLE
Feature request: JUnit format output option.

### ISSUE THREAD

**Issue by bcsgh** [priv-kweihmann/oelint-adv#888]
It would be nice to have the option of passing `--format=junit`.

When used in automation (e.g. CI) it's often desirable to have machine readable output formats. One of the formats more commonly supported by consumers is the JUnit XML format. (A number of other linters I'm already using offer this as an option.)

**Comment by priv-kweihmann** [comment]
I started a rough implementation of a specific format output option.
Sadly Junit is not a very well defined standard, so what I came up with would look like this

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
    <testsuite id="oelint-adv" name="oelint-adv" tests="19" failures="19" errors="0" skipped="0">
        <testcase name="oelint.bbclass.underscores" file="/path/to/layer/meta/classes-recipe/image-artifact-names.bbclass" line="1">
            <failure message="oelint.bbclass.underscores" type="failure">
                /path/to/layer/meta/classes-recipe/image-artifact-names.bbclass:1:error:oelint.bbclass.underscores:bbclass filenames should not contain '-'. Replace it by '_' [branch:true]</failure>
        </testcase>
        <testcase name="oelint.task.noanonpython" file="/path/to/layer/meta/classes-recipe/image-artifact-names.bbclass" line="36">
            <failure message="oelint.task.noanonpython" type="failure">
                /path/to/layer/meta/classes-recipe/image-artifact-names.bbclass:36:warning:oelint.task.noanonpython:Avoid anonymous python functions as they are expensive and come with all sorts of side effects [branch:true]</failure>
        </testcase>
        <testcase name="oelint.bbclass.underscores" file="/path/to/layer/meta/classes-recipe/image-postinst-intercepts.bbclass" line="1">
            <failure message="oelint.bbclass.underscores" type="failure">
                /path/to/layer/meta/classes-recipe/image-postinst-intercepts.bbclass:1:error:oelint.bbclass.underscores:bbclass filenames should not contain '-'. Replace it by '_' [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.specific" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="63">
            <failure message="oelint.vars.specific" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:63:error:oelint.vars.specific:'SDK_TOOLCHAIN_LANGS' is set specific to ['sdkmingw32'], but isn't known from PACKAGES, MACHINE, DISTRO or resources [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.specific" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="67">
            <failure message="oelint.vars.specific" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:67:error:oelint.vars.specific:'SDK_TOOLCHAIN_LANGS' is set specific to ['linux-gnux32'], but isn't known from PACKAGES, MACHINE, DISTRO or resources [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.specific" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="69">
            <failure message="oelint.vars.specific" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:69:error:oelint.vars.specific:'SDK_TOOLCHAIN_LANGS' is set specific to ['sdkmingw32'], but isn't known from PACKAGES, MACHINE, DISTRO or resources [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.notneededspace" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="72">
            <failure message="oelint.vars.notneededspace" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:72:info:oelint.vars.notneededspace:Space at the beginning of the var is not needed [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.notneededspace" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="79">
            <failure message="oelint.vars.notneededspace" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:79:info:oelint.vars.notneededspace:Space at the beginning of the var is not needed [branch:true]</failure>
        </testcase>
        <testcase name="oelint.task.noanonpython" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="98">
            <failure message="oelint.task.noanonpython" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:98:warning:oelint.task.noanonpython:Avoid anonymous python functions as they are expensive and come with all sorts of side effects [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.spacesassignment" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="183">
            <failure message="oelint.vars.spacesassignment" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:183:warning:oelint.vars.spacesassignment:Suggest spaces around assignment. E.g. 'FOO = "BAR"' [branch:true]</failure>
        </testcase>
        <testcase name="oelint.spaces.lineend" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="251">
            <failure message="oelint.spaces.lineend" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:251:warning:oelint.spaces.lineend:Line shall not end with a space [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.mispell.unknown" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="264">
            <failure message="oelint.vars.mispell.unknown" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:264:warning:oelint.vars.mispell.unknown:'SSTATE_SKIP_CREATION' is unknown. Consider adding it via --constantmods [branch:true]</failure>
        </testcase>
        <testcase name="oelint.tabs.notabs" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="274">
            <failure message="oelint.tabs.notabs" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:274:warning:oelint.tabs.notabs:Don't use tabs use spaces [branch:true]</failure>
        </testcase>
        <testcase name="oelint.vars.pathhardcode.sysconfdir" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="274">
            <failure message="oelint.vars.pathhardcode.sysconfdir" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:274:warning:oelint.vars.pathhardcode.sysconfdir:'${sysconfdir}' should be used instead of '/etc' [branch:true]</failure>
        </testcase>
        <testcase name="oelint.tabs.notabs" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="342">
            <failure message="oelint.tabs.notabs" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:342:warning:oelint.tabs.notabs:Don't use tabs use spaces [branch:true]</failure>
        </testcase>
        <testcase name="oelint.tabs.notabs" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="351">
            <failure message="oelint.tabs.notabs" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:351:warning:oelint.tabs.notabs:Don't use tabs use spaces [branch:true]</failure>
        </testcase>
        <testcase name="oelint.tabs.notabs" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="395">
            <failure message="oelint.tabs.notabs" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:395:warning:oelint.tabs.notabs:Don't use tabs use spaces [branch:true]</failure>
        </testcase>
        <testcase name="oelint.task.multifragments" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="423">
            <failure message="oelint.task.multifragments" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:423:info:oelint.task.multifragments:Multiple fragments of the same function in the same file should be merged [branch:true]</failure>
        </testcase>
        <testcase name="oelint.task.noanonpython" file="/path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass" line="423">
            <failure message="oelint.task.noanonpython" type="failure">
                /path/to/layer/meta/classes-recipe/populate_sdk_base.bbclass:423:warning:oelint.task.noanonpython:Avoid anonymous python functions as they are expensive and come with all sorts of side effects [branch:true]</failure>
        </testcase>
    </testsuite>
</testsuites>
```

It passes https://lotterfriends.github.io/online-junit-parser/, but I can't tell if your tool suite can deal with the output.
Please use above's output and tell me if that fits your toolchain or what other mandatory fields are required

**Comment by priv-kweihmann** [comment]
In general the number of tests overall can't be computed by the tool I'm afraid, so either you will have 1 test overall, 0 failed for a good run or number of issues for both overall test count and failed tests.
I know that might turn into weird looking graphs in e.g. Jenkins, but to me that would be acceptable

**Comment by bcsgh** [comment]
I'd have to check what other linters do, but I'd think either of reporting zero tests passing/failing on success or inserting just a passing "success test" when nothing else is there would be the first two choices. Another option would be to always include a summary test "per unit" (e.g. one per file or per "whatever makes sense".)

As for format: the consumer I care about is Jenkins (and it will be a while before I testing this gets through the "what I work on today" queue, I wasn't expecting this fast a response; thanks!).

**Comment by bcsgh** [comment]
On the prior topic, a "synthetic success" testcase would likely be better than no testcases; there are a few tools where "no passing tests" is treated as an error even when there are no failing tests (e.g. to flag when tests don't run at all).

**Comment by priv-kweihmann** [comment]
`1 test overall, 0 failed` should imply that. I'm aware of esp Jenkins being picky on such input files, so in the end the pseudo test will be mapped to be a success
